### PR TITLE
(Towards #717) refactor functionality that determines rank of array references.

### LINKED ISSUE #717: Where statement array syntax is limited
Array syntax in Fortran where statements must be for the full extent i.e. the dimension must only contain a `:`. If it is for a subset of the data in the dimension e.g. `2:4` then the PSyIR front-end raises an exception.
We also do not currently support use of 'bare' arrays, e.g. `WHERE(some_array < 0)` because we rely on array notation to identify array accesses. Once #500 is complete we will have the necessary functionality to look-up the shape of such arrays.

### PR DESCRIPTION
**Pull request by arporter**

### CODE REVIEW

**Comment by arporter** [comment]
This is ready for a first look now, probably from @sergisiso. In the end it just rationalises what was duplicated functionality to find the rank of an array subsection. It may be that we really want the routine (rank_of_subsection) to give us back the *shape* of any subsection, possibly using `Range` nodes.  I could change what I have to do that or we could stick with what I have so far and do that as a subsequent step.

**Review by sergisiso — CHANGES_REQUESTED** [review]
You maybe already hinted it here:

> It may be that we really want the routine (rank_of_subsection) to give us back the shape of any subsection

But this recursion method to find the slice rank is somewhat similar to what @rupertford was planning to implement in order to 'bubble up' the datatypes too all PSyIR expression (datanodes) nodes. I can't find the specific issue but we had a chat during Easter week and Rupert was about to started implementing it. (starting with operations but accessors must also be part of it for a complete solution)

Given that this .datatype property for all expression nodes will be more generic (and would avoid having the slightly confusing "rank_of_subsection" in nodes that have nothing to do with arrays like member and structurereference) I think the methods you add in this PR could at some point be superseded by .datatype and simply be: 
```
rank_slice = len(top_node.datatype.shape)
```

Note that for a reference `a(3,:)`
`ref.symbol.datatype is ArrayType([3, Extent.ATTRIBUTE])`
but
`ref.datatype is  ArrayType([Extent.ATTRIBUTE])`
which is the information you need. 

Since you may need this to fix current Nemo issues with where statements and the .datatype may take a while. I am fine to accept this PR if you address the other comments I have and add some comments/TODO about the possible temporary nature of the methods.
But if @rupertford can shed some light on the progress on this and maybe you can combine efforts to go for a single solution, it would be even better.

File: src/psyclone/psyir/nodes/structure_reference.py [hunk 1]
```diff
@@ -189,6 +197,15 @@ def _create(cls, symbol, symbol_type, members, parent=None):
         for component in reversed(members[:-1]):
             if isinstance(component, tuple):
                 # This is an array access so we have an ArrayOfStructuresMember
+                if any(isinstance(idx_expr, Range) for idx_expr in
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyir/nodes/structure_reference.py:197]: Yes, my understanding is that it is not valid anywhere.

File: src/psyclone/tests/psyir/nodes/array_member_test.py [hunk 2]
```diff
@@ -39,7 +39,7 @@
 from __future__ import absolute_import
 import pytest
 from psyclone.psyir import symbols, nodes
-from psyclone.errors import GenerationError
+from psyclone.errors import GenerationError, InternalError
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/tests/psyir/nodes/array_member_test.py:42]: Fixed.

**Pull request closed without merging**